internal/layer2/announcer.go

```diff
@@ -131,24 +137,49 @@ func (a *Announce) updateInterfaces() {
 	return
 }
 
-func (a *Announce) spam(name string) {
-	// TODO: should abort if we lose control of the IP mid-spam.
-	start := time.Now()
-	for time.Since(start) < 5*time.Second {
-		if err := a.gratuitous(name); err != nil {
-			a.logger.Log("op", "gratuitousAnnounce", "error", err, "service", name, "msg", "failed to make gratuitous IP announcement")
+func (a *Announce) spamLoop() {
+	m := map[string]time.Time{}
+	// see https://github.com/metallb/metallb/issues/172 for the 1100 choice
+	ticker := time.NewTicker(1100 * time.Millisecond)
+	for {
+		select {
+		case ip := <-a.spamCh:
+			ips := ip.String()
+			_, ok := m[ips]
+			m[ips] = time.Now().Add(5 * time.Second)
+			if !ok {
+				// The first 2 spam can be close
+				// but we don't want to wait 1.1 sec
```

Comment:
- I don't understand this comment. Could you please make it clearer?

Reply:
Is it better ?
```
// The first 2 calls to spam() can be msec away,
// but we can't wait 1.1 sec
```

Reply:
I still don't understand the technical meaning behind "but we can't wait 1.1 sec" :confused: What are we trying to say here?

Anyway, let's not block on this. I'd just change `msec` to `milliseconds` and add a period at the end for consistency.

Reply:
If the ticker is already running (always the case right now), we might call spam() because it's a new IP, and 1msec after call spam() again in the ticker part. This is not ideal, but the worst case, where we wait 1099 ms for the first spam() call is not acceptable.

Reply:
I *think* I understand what you mean now. Can I rephrase the comment to the following?

```
				// Spam right away to avoid waiting 1100 milliseconds even if
				// it means we call spam() twice in a row in a short amount of
				// time.
```

Reply:
Done
```
// Spam right away to avoid waiting up to 1100 milliseconds even if
// it means we call spam() twice in a row in a short amount of time.
```